### PR TITLE
Create one normal distribution per omp thread to avoid data races

### PR DESCRIPTION
**Pull request by markus-jehl**
This addresses the second part of https://github.com/UCL/STIR/issues/1255

### CODE REVIEW

File: src/include/stir/DetectorCoordinateMap.h [hunk 1]
```diff
@@ -27,6 +27,7 @@
 #include <map>
 #include <boost/algorithm/string.hpp>
 #include <boost/unordered_map.hpp>
+#include <omp.h>
```
Comment (KrisThielemans) [inline src/include/stir/DetectorCoordinateMap.h:30]:
```suggestion
#ifdef STIR_OPENMP
#include <omp.h>
#endif
```

File: src/include/stir/DetectorCoordinateMap.h [hunk 2]
```diff
@@ -65,14 +66,32 @@ class DetectorCoordinateMap
...
+		{ 
+      read_detectormap_from_file(filename);
+#ifdef STIR_OPENMP
+      distributions.resize(omp_get_max_threads());
+      for (auto &distribution : distributions)
+        distribution = std::normal_distribution<double>(0.0, sigma);
+#else
+      distributions.resize(1);
+      distributions[0] = std::normal_distribution<double>(0.0, sigma);
+#endif
```
Comment (KrisThielemans) [inline src/include/stir/DetectorCoordinateMap.h:79]:
```suggestion
		DetectorCoordinateMap(sigma)
		{ 
                   read_detectormap_from_file(filename);
```
should work I believe (delegating constructor)

File: src/include/stir/DetectorCoordinateMap.h [hunk 3]
```diff
@@ -88,9 +107,15 @@ class DetectorCoordinateMap
...
-		coord.z() += static_cast<float>(distribution(generator));
+#ifdef STIR_OPENMP
+    auto thread_id = omp_get_thread_num();
+#else
+    auto thread_id = 0;
+#endif
+    
+		coord.x() += static_cast<float>(distributions[thread_id](generator));
+		coord.y() += static_cast<float>(distributions[thread_id](generator));
+		coord.z() += static_cast<float>(distributions[thread_id](generator));
```
Comment (KrisThielemans) [inline src/include/stir/DetectorCoordinateMap.h:118]:
```suggestion
                auto& distribution = distributions[thread_id];
		coord.x() += static_cast<float>(distribution(generator));
		coord.y() += static_cast<float>(distribution(generator));
		coord.z() += static_cast<float>(distribution(generator));
```

However, this raises the question if `generator` is thread-safe. I have no idea, but somehow I think it's that one that isn't, while the distribution itself is. Not sure!

File: src/include/stir/DetectorCoordinateMap.h [hunk 4]
```diff
@@ -88,9 +107,15 @@ class DetectorCoordinateMap
 	stir::CartesianCoordinate3D<float> get_coordinate_for_det_pos( const stir::DetectionPosition<>& det_pos ) const
 	{ 
 		auto coord = det_pos_to_coord.at(det_pos);
-		coord.x() += static_cast<float>(distribution(generator));
-		coord.y() += static_cast<float>(distribution(generator));
-		coord.z() += static_cast<float>(distribution(generator));
+#ifdef STIR_OPENMP
+    auto thread_id = omp_get_thread_num();
+#else
+    auto thread_id = 0;
```
Comment (KrisThielemans) [inline src/include/stir/DetectorCoordinateMap.h:113]: more white-space here apparently (unless it's GitHub).

**Review by KrisThielemans — COMMENTED** [review]
cosmetics but one question where the thread-safety should sit: the distribution or the generator?

File: src/include/stir/DetectorCoordinateMap.h (hunk 2 above)
Comment (markus-jehl) [reply src/include/stir/DetectorCoordinateMap.h:79]: True, that's much cleaner!

File: src/include/stir/DetectorCoordinateMap.h (hunk 3 above)
Comment (markus-jehl) [reply src/include/stir/DetectorCoordinateMap.h:118]: good that you asked! I initially thought it was just the distribution (and the intel inspector complaints have gone away), but some more research showed that the generator is also not thread-safe. I'll just store a vector of instances for each.

**Review by KrisThielemans — CHANGES_REQUESTED** [review]
I think this is fine, aside from the `explicit`.

By the way, would it be faster to have `if(sigma ==0) {return coord;}` without going throught the random generator?

File: src/include/stir/DetectorCoordinateMap.h [hunk 5]
```diff
@@ -113,9 +124,19 @@ class DetectorCoordinateMap
 protected:
 
  DetectorCoordinateMap(double sigma = 0.0) :
```
[the remark this answers is not in the document]
Comment (markus-jehl) [reply src/include/stir/DetectorCoordinateMap.h:126]: good idea! it can't hurt to handle the sigma == 0.0 case separately.

**Review by KrisThielemans — APPROVED** [review]

**Pull request merged**